File: CodingInterviewsJavaSolutions/Minimum Area Rectangle.java
class Solution {
    public int minAreaRect(int[][] points) {
        HashMap<Integer, HashSet<Integer>> xToYs = new HashMap<>();
        int sz = points.length;
        int ans = Integer.MAX_VALUE;

        for (int i = 0; i < sz; i++) {
            int x = points[i][0];
            int y = points[i][1];
            xToYs.computeIfAbsent(x, k -> new HashSet<>()).add(y);
        }

        for (int i = 0; i < sz; i++) {
            int x1 = points[i][0];
            int y1 = points[i][1];

            for (int j = i + 1; j < sz; j++) {
                int x2 = points[j][0];
                int y2 = points[j][1];

                if (x1 == x2 || y1 == y2)
                    continue;

                if (!xToYs.get(x1).contains(y2) || !xToYs.get(x2).contains(y1))
                    continue;

                int area = Math.abs(x2 - x1) * Math.abs(y2 - y1);
                ans = Math.min(ans, area);
            }
        }

        if (ans == Integer.MAX_VALUE)
            return 0;
        return ans;
    }
}

File: CodingInterviewsPythonSolutions/Minimum Area Rectangle.py
class Solution:
    def minAreaRect(self, points: List[List[int]]) -> int:
        x_to_ys = {}
        sz = len(points)
        ans = float('inf')

        for i in range(sz):
            x = points[i][0]
            y = points[i][1]
            if x not in x_to_ys:
                x_to_ys[x] = set()
            x_to_ys[x].add(y)

        for i in range(sz):
            x1 = points[i][0]
            y1 = points[i][1]

            for j in range(i + 1, sz):
                x2 = points[j][0]
                y2 = points[j][1]

                if x1 == x2 or y1 == y2:
                    continue

                if y2 not in x_to_ys[x1] or y1 not in x_to_ys[x2]:
                    continue

                area = abs(x2 - x1) * abs(y2 - y1)
                ans = min(ans, area)

        if ans == float('inf'):
            return 0
        return ans
